Remove Renderer3 and renderComponent

alxhub opened the pull request:
See individual commits

alxhub committed:
refactor(core): remove experimental renderComponent function

This commit removes the experimental `renderComponent` operation, which
implemented an alternative, lighter-weight bootstrapping option. This
experiment is being discontinued as the path towards it being a supported
production API was not clear. In particular, it lacked a clear roadmap for
supporting Angular's change detection, particularly when consuming existing
components.

alxhub committed:
test(core): migrate tests off of Renderer3 interfaces

This commit migrates any remaining Angular tests which are using some form
of Renderer3 interfaces. Instead, they're switched to Renderer2.

alxhub committed:
refactor(core): disable automatic usage of Renderer3

Previously, when instantiating a component, Angular would look in the DI
hierarchy for `RendererFactory2`. Any DI tree which rolls up through an
application injector (that is, one created with `BrowserModule`) should be
able to provide this interface. If not found, Angular would switch to the
experimental `Renderer3` mechanism. This switch was designed this way,
because it allowed for the creation of experimental applications where
`RendererFactory2` was not included in the bundle at all.

In this commit, instead of automatically falling back on `Renderer3`-style
rendering, an error is raised instead if `RendererFactory2` is missing from
the DI hierarchy.

alxhub committed:
refactor(core): remove experimental `Renderer3` abstraction

This commit removes the `Renderer3` experiment which attempted to use the
real DOM API as Angular's renderer. As shown in the diff, having this
experiment around added real code complexity to Angular that could not be
removed by an optimizer.

Since we no longer feel this experiment is worth continuing, we're removing
the `Renderer3` concept and all supporting code.

alfaproject commented:
OOC was it for performance reasons you decided to drop renderer 3, or? What was the goal/experiment in the first place? I've tried to search for my own answer but couldn't find, sorry if it's written somewhere already.

alxhub committed:
refactor(core): rename the ProceduralRenderer3 interface to Renderer

Due to the restrictions of circular dependency checking, we need a separate
internal interface for a renderer. We cannot use `Renderer2` as a type
internally as even importing it as a type incurs a dependency on its
implementation, which creates a major potential for circular dependencies.

Previously this role was served by the `Renderer3` type. As we prepare to
remove the `Renderer3` abstraction (the idea of using `document` as a
differently-shaped renderer), this commit renames `ProceduralRenderer3` to
the more generic term `Renderer`.

`RendererFactory3` is also renamed to the more generic `RendererFactory` for
consistency.

alxhub commented:
@alfaproject Angular has historically had a "Renderer" abstraction (currently, the interface `Renderer2`). This adds overhead over direct DOM manipulation, but gives a number of benefits (it's a natural layer, for example, to implement CSS encapsulation and animation of added/removed elements). Theoretically, an application could swap out the underlying DOM entirely, and use the framework to render in a completely different context (like NativeScript for example).

`Renderer3` was a somewhat successful experiment to remove the overhead of the renderer abstraction by using the DOM itself directly. While this did work and was faster, it had several challenges:

1. As mentioned above, the renderer layer is used for things like CSS encapsulation and animations. It was unclear how users who needed to directly manipulate DOM should do so in a way that was compatible with these features, without the Renderer abstraction.

2. Having both codepaths (the production Renderer2 path, and the experimental path) in the runtime adds non-tree-shakable code to support the unused experiment, as well as complexity that anyone working on runtime code would need to navigate.

3. Many of our tests were using the experimental rendering approach, which meant they were not testing "reality" - the test configuration looked different than that of a real application. This could lead to bugs.

4. Even though the direct-to-DOM rendering system was intended to be only an internal experiment, the way it was implemented meant that production applications could accidentally instantiate a component in such a way as to activated. There was at least one case of this happening within Google, causing some very strange errors that were very difficult for the engineer who stumbled over this to debug.

So we're removing this experiment, because we've learned everything we needed to from it, we don't see a way to easily ship it, and it's costing us too much to keep the code supporting it in place.

AndrewKushnir commented on packages/core/src/render3/instructions/styling.ts:
```diff
@@ -786,7 +787,8 @@ function findStylingValue(
     }
     let currentValue = isStylingMap ? keyValueArrayGet(valueAtLViewIndex, prop) :
-                                      key === prop ? valueAtLViewIndex : undefined;
+        key === prop                ? valueAtLViewIndex :
```
The formatting looks very unusual... May be wrap internal expression into `(...)`?

alxhub replied:
Lol, no kidding. The formatter really gave up here.

AndrewKushnir reviewed (approved):
@alxhub looks great 👍

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: size-tracking, public-api

pkozlowski-opensource reviewed (approved):
Yay! 🎉

alfaproject commented:
@alxhub thank you for the detailed answer. Today I learned something new (:

atscott commented:
This PR was merged into the repository by commit 8d2e5e6c4f5dd8ff435152f4a3ce94d7197255aa.
